[레베카] 로또 미션 제출합니다.

chws opened the pull request:
리뷰 항상 감사합니다!
잘부탁드립니다:)

LTTTTTE committed:
docs: 기능 목록 작성

LTTTTTE committed:
feat: 구매한 로또갯수 계산하는 기능 추가

LTTTTTE committed:
feat: 랜덤수 생성하는 기능 추가

LTTTTTE committed:
feat: 수량에 맞는 로또를 생성하는 기능 추가

LTTTTTE committed:
feat: 구입 금액이 자연수인지 확인하는 기능

LTTTTTE committed:
test: Money domain initTestCode 추가

LTTTTTE committed:
feat: InputView 클래스 추가

LTTTTTE committed:
feat: 지난주 당첨번호 parse 하는 기능 추가

LTTTTTE committed:
feat: 당첨 번호의 예외 처리 기능 추가

LTTTTTE committed:
feat: 로또 번호와 보너스 번호의 중복 체크 기능 추가

LTTTTTE committed:
test: 로또 몇 등인지 확인하는 테스트 추가

LTTTTTE committed:
feat: 각 로또가 당첨번호/보너스번호 중 몇 개 일치하는지 확인하는 기능 추가

LTTTTTE committed:
feat: 당첨 통계 출력하는 기능 추가

LTTTTTE committed:
feat: 구매한 로또갯수 출력하는 기능 추가

LTTTTTE committed:
feat: 구입한 로또 출력하는 기능 추가

LTTTTTE committed:
feat: 사용자로부터 입력받는 기능 추가

LTTTTTE committed:
refactor: 구입한 로또가 정렬될수 있도록 수정

LTTTTTE committed:
refactor: 범용성을 가지도록 메소드 명 수정

LTTTTTE committed:
feat: 등수에 없는 Rank 처리하는 기능 추가

LTTTTTE committed:
feat: 메인 컨트롤러 작성

LTTTTTE committed:
test: Lotto 객체 생성 테스트 코드 추가

LTTTTTE committed:
test: NumberParser 테스트코드 작성

LTTTTTE committed:
test: BonusNumber 테스트 작성

LTTTTTE committed:
test: WinningNumber 테스트 코드 작성

LTTTTTE committed:
refactor: 패키지 분리

LTTTTTE committed:
refactor: 원시값 포장

LTTTTTE committed:
feat: 커스텀 예외처리 추가

LTTTTTE committed:
refactor: 모든 문자열 포장

LTTTTTE committed:
refactor: 메인컨트롤러 기능 분리

chws committed:
Merge pull request #1 from LTTTTTE/oguRebecca

페어 코드 분리

jihan805 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+		this.numbers = numbers;
+	}
+
+	private boolean hasDuplicatedNumbers(List<Integer> numbers) {
```
List 자료구조가 아닌 Set 자료구조를 활용하는 것은 어떨까요?
(Set을 활용했을 때 얻을 수 있는 이점)

chws replied:
중복을 제거하기 위해서 리뷰해주신대로 Set을 이용해보았습니다. :) 
그런데 로또의 숫자 배열 자체를 Set으로 갖지 않고 로또 생성자에서 중복을 거를 때에만 Set을 이용해도 효과적인 것 같아서, 처음에는 전체적으로 Set으로 바꾸려다가 그 부분에서만 Set을 이용했습니다. 
이 부분에 대해서 피드백 주시면 감사하겠습니다~!

jihan805 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,28 @@
+		return money.getMoney() / LOTTO_PRICE;
+	}
+
+	public static Lotto generateSingleLotto() {
```
1. 사용하는 메서드 아래 위치 (메서드 선언 순서)
2. public으로 선언해야할까요? 

위 두가지 피드백에 대해 코드 전체적으로 검토해보고 적용해보면 어떨까요?

jihan805 commented on src/main/java/lotto/controller/RankCalculator.java:
```diff
@@ -0,0 +1,36 @@
+
+public class RankCalculator {
+	public static List<Rank> calculateMultiple(List<Lotto> lottos, WinningNumber winningNumber) {
+		final List<Integer> winningNumbers = winningNumber.getWinningNumber().getNumbers();
```
규칙 4: 한 줄에 점을 하나만 찍는다. - 위반

jihan805 commented on src/main/java/lotto/controller/RankCalculator.java:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	private static boolean hasBonusNumber(Lotto lotto, int bonusNumber) {
+		return lotto.getNumbers().contains(bonusNumber);
```
get하는 대신, lotto.hasBonusNumber(bonusNumber) 와 같이 구현할 수 있지 않을까요?

jihan805 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,28 @@
+public class LottoManager {
+	private static final int LOTTO_PRICE = 1000;
+	private static final LottoNumberGenerator lottoNumberGenerator = new LottoNumberGenerator();
+
```
돈을 나타내는 숫자의 경우 천 단위를 구분하기 위해 "1_000"와 같이!

jihan805 commented on src/main/java/lotto/utils/NumberGenerator.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+
+public interface NumberGenerator {
+	List<Integer> generate();
```
인터페이스 추출 👍 
다음단계에서 잘 활용해보면 좋을 것 같아요.

jihan805 commented on src/test/java/lotto/domain/BonusNumberTest.java:
```diff
@@ -0,0 +1,18 @@
+		assertThatThrownBy(() -> new BonusNumber(46))
+			.isInstanceOf(LottoNumberRangeException.class)
+		.hasMessageContaining("보너스 넘버");
+	}
```
하나의 테스트에서 두가지를 테스트하고 있는 것 같아요!
initWinningNumberTest()도 마찬가지구요!

jihan805 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,65 @@
+			.filter(Objects::nonNull)
+			.mapToInt(Rank::getReward)
+			.sum();
+		System.out.printf(EARNING_RATE_MESSAGE, (double)sum / money.getMoney() * 100);
```
수익률에 대한 계산을 OutputView 에서 하고 있어요!
OutputView의 역할로 수익률 계산이 적절할까요?

jihan805 reviewed (changes requested):
안녕하세요 ~ 깔끔하게 구현 💯 
간단한 피드백 남겼으니 확인 후 다음단계 진행하면 좋을 것 같아요 :)

Commit:
refactor: 리뷰 바탕으로 수정

 - 중복된 로또인지 판별할 때 Set 이용
 - 메소드가 필요없이 public일 때 private으로 바꾸기
 - 메소드 단순화
 - 수익률 계산 AccountingManager에서 하도록 변경
 - 테스트 분리

jihan805 commented on src/main/java/lotto/controller/AccountingManager.java:
```diff
@@ -0,0 +1,17 @@
+                .filter(Objects::nonNull)
+                .mapToInt(Rank::getReward)
+                .sum();
+        return (double) sum / money.getMoney() * 100;
```
상수 사용 !

jihan805 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,46 @@
+    private static final String LOTTO_NUMBER_RANGE_MESSAGE = "1~45 범위의 숫자만 로또 번호가 될 수 있습니다.";
+    private static final String LOTTO_NUMBER_DUPLICATED_MESSAGE = "잘못된 로또 번호입니다. 중복 안됨, 갯수는 6개";
+
+    private List<Integer> numbers;
```
1. 숫자 값 하나를 추상화한 LottoNo/LottoNumber와 같은 객체를 추가하는 것은 어떨까요?
위의 객체의 생성자에서 1 ~ 45의 값인지에 대한 유효성 체크도 할 수 있지 않을까요?
BonusNumber 클래스를 조금 바꾸어 활용하여도 좋을 것 같아요 :) 

2. 아래와 같이 검증을 위해 Set을 사용하는 것, 변수 선언 자체를 Set을 하는 것 두 방법 모두 괜찮다고 생각합니다!

jihan805 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private static void showStatistics(List<Rank> ranks) {
+        Map<Rank, Long> rankCounts = ranks.stream()
```
1. GameResult와 같은 클래스로 추출하는 것이 더 적절하지 않을까요?
이 역시 OutputView의 역할로는 맞지 않는 것 같아요 :) 
2. 메서드 선언 순서 다시 확인해주세요!
3. 수익률 등 결과와 관련된 테스트가 보이지 않네요 ㅠ 해당 메서드를 추출하고 나면 테스트 작성이 쉬워질 수 있지 않을까요?
(Money 클래스와, 게임 결과를 구하는 클래스 등을 활용해보면 어떨까요?)

테스트는 미션 진행중 가장 중요한 것 중 하나라고 생각합니다! 다른 부분 구현을 너무 잘해주셨지만, 위 내용에 대해  한번 더 고민해보고 다음 단계 코드를 구현하는 것이 더 좋을 것 같습니다.

다음 단계 진행 전 위 내용에 대해 먼저 반영 후 진행해주세요! 
커밋 로그를 통해 확인할 수 있게 커밋을 나누어 작성해주세요  :)

jihan805 commented on src/test/java/lotto/domain/MoneyTest.java:
```diff
@@ -0,0 +1,30 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class MoneyTest {
```
정상적인 경우에 대한 테스트는 생략해도 되는걸까요?

jihan805 reviewed (approved):
안녕하세요~! 피드백 반영 & 전체적인 구현 💯 
간단한 피드백 추가로 남겼으니 확인후 다음단계 진행해 주세요 :)
